KOGITO-6854 Added create-your-first-workflow-service guide

hbelmiro opened the pull request:
JIRA: https://issues.redhat.com/browse/KOGITO-6854
This PR depends on: https://github.com/kiegroup/kogito-examples/pull/1266

hbelmiro commented:
@ricardozanini can you please review this PR?

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,274 @@
+
+[source,shell]
+----
+quarkus create app -x=kogito-quarkus-serverless-workflow,quarkus-container-image-jib,quarkus-resteasy-jackson,quarkus-smallrye-openapi org.kie.kogito.examples:serverless-workflow-hello-world:1.0
```
What about providing also the quarkus-maven-plugin command? Then we have to care about one prerequisite less.

ricardozanini replied:
Please also separate this command in separated lines with `\`:

```suggestion
quarkus create app \
    -x=kogito-quarkus-serverless-workflow \
    -x=quarkus-container-image-jib \
    -x=quarkus-resteasy-jackson \
    -x=quarkus-smallrye-openapi \
    org.kie.kogito.examples:serverless-workflow-hello-world:1.0
```

Also, please explain what we are doing here. This way, newcomers will understand what they are doing rather than just copying and pasting.

hbelmiro commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,274 @@
+}
+----
+
+Before running the test, run the following command to build the application:
+[source,shell]
+----
+./mvnw package -DskipTests
+----
```
@ricardozanini do you know if there's a way to remove this step?

ricardozanini replied:
Since it's an integration test, we can't. But I think you can do this test using the regular `QuarkusTest`? https://quarkus.io/guides/getting-started-testing#recap-of-http-based-testing-in-jvm-mode

Also, please add the reference to all Quarkus Guides, so people can better understand where to go in case of problems.

hbelmiro replied:
> Since it's an integration test, we can't. But I think you can do this test using the regular QuarkusTest?

Yes, we can. It was not working before, but now it does. I might have done something stupid.
Now I can remove that annoying step :)

MarianMacik commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,274 @@
+ --/ __ \/ / / / _ | / _ \/ //_/ / / / __/
+ -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \
+--\___\_\____/_/ |_/_/|_/_/|_|\____/___/
+2022-05-25 14:38:09,741 INFO  [org.kie.kog.add.qua.mes.com.QuarkusKogitoExtensionInitializer] (Quarkus Main Thread) Registered Kogito CloudEvent extension
```
This log can be shortened a bit maybe?

MarianMacik reviewed (commented):
Looks good, just a couple of recommendations.

akumar074 commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,274 @@
+However, you can go right to the completed example by cloning the git repository:
+
+The solution is located in the `kogito-quarkus-examples/serverless-workflow-hello-world`
+https://github.com/hbelmiro/kogito-examples/tree/main/kogito-quarkus-examples/serverless-workflow-hello-world[directory].
```
This link is broken and redirects to a downstream repo. This should refer to upstream `kogito-example` repo.

hbelmiro replied:
Good catch regarding the downstream repo :)
The link is broken because it still doesn't exist. It's supposed to be created in https://github.com/kiegroup/kogito-examples/pull/1266, which is a dependency of this PR.
Anyway, the link is going to change because a new repo will be created for the guides.

akumar074 reviewed (commented):
Hello @hbelmiro Thanks for the PR. Please find my comment below.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,274 @@
+* https://maven.apache.org/install.html[Apache Maven]
+* https://quarkus.io/guides/cli-tooling[Quarkus CLI]
+
+To edit your workflows:
```
It's better to link to the tooling guide: https://github.com/kiegroup/kogito-docs/blob/main/serverlessworkflow/modules/ROOT/pages/getting-started/getting-familiar-with-our-tooling.adoc

ricardozanini replied:
In this tooling tutorial, we should introduce Maven and Quarkus CLI since they are part of the developer toolkit.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,274 @@
+
+You can also run the test from your IDE directly (be sure you stopped the application first).
+
+== What's next?
```
People can also go to the Cloud category from here to understand how to deploy it in the cloud. cc @amadhusu

ricardozanini reviewed (commented):
Looks great! Just a few comments, if you don't mind. :)

hbelmiro commented:
Thanks for your review @akumar074, @ricardozanini, @MarianMacik.
I pushed the requested changes.

fjtirado commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,255 @@
+* `kogito-quarkus-serverless-workflow` - Adds support for Serverless Workflow
+* `quarkus-container-image-jib` - Adds support for Container Image Builds
+* `quarkus-resteasy-reactive-jackson` - Adds support for Reactive RESTEasy
+* `quarkus-smallrye-openapi` - Adds support for OpenAPI
```
I will clarify that this dependecy is needed for support generation of Swagger documentation. Just Open API might lead to confusion, since SWF use OpenAPI spec to invoke remote REST servers, but the guide does not include any (thats good)

fjtirado commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,255 @@
+
+* `kogito-quarkus-serverless-workflow` - Adds support for Serverless Workflow
+* `quarkus-container-image-jib` - Adds support for Container Image Builds
+* `quarkus-resteasy-reactive-jackson` - Adds support for Reactive RESTEasy
```
I think we need to add here hat this dependency is needed for the generated REST resources that are used to start the flow process throug an HTTP invocation

fjtirado reviewed (approved):
Great work @hbelmiro 
Just a couple of minor comments. Basically to clarify the purpose of the "default" dependencies.

hbelmiro commented:
> Great work @hbelmiro Just a couple of minor comments. Basically to clarify the purpose of the "default" dependencies.

Many thanks for the feedback, @fjtirado.
I pushed the suggested changes.

akumar074 reviewed (approved):
@hbelmiro This looks good to me. Thanks!

ibek commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,257 @@
+* Running the application
+* Testing the application
+
+== Prerequisites
```
Shouldn't there be prerequisites such as:

time needed to go through the guide? 10 minutes?
JDK 11?
Apache Maven minimal version?
Optionally Quarkus CLI

hbelmiro replied:
Linking to the guide was a suggestion by @ricardozanini [here](https://github.com/kiegroup/kogito-docs/pull/2#discussion_r882846710).

ibek replied:
@ricardozanini  I think this guide should work as a starting point without the need to go to another guide. JDK 11, Maven should be the minimum in my opinion. We could say that if they are interested in more tooling they can follow the tooling guide as you suggest. WDYT?

ibek commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,257 @@
+    }
+  ]
+}
+----
```
I believe the workflow definition would deserve some description and guidance on how to read it. That, workflow definition is uniquely identified by id and generated REST endpoints will be based on that. Also that the workflow contains 2 states. The one where workflow starts (start) is "Inject Hello World" of the type Inject - perhaps add some definition of Inject state from the spec what it does - in our case it sets greeting field name in workflow data. Then transitions to "Inject Mantra" where it sets "mantra" field name in workflow data and that's where the workflow ends. The result/output of workflow is the workflow data. --- or something like that.

ibek commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,257 @@
+{"id":"efb59bfa-ad9c-4062-a6d2-2d9184dd4b3d","workflowdata":{"greeting":"Hello World","mantra":"Serverless Workflow is awesome!"}}
+----
+
+Hit `CTRL+C` to stop the application, or keep it running and enjoy the blazing fast hot-reload.
```
Before stopping the application ask the reader to update the workflow (greeting and/or mantra) and resend the request to see the changed workflow results. I believe we should encourage readers to play and learn. Then we can explain that running in dev mode has the advantage of hot-reload, etc. Afterwards we can put the command to stop the application.

ibek commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,257 @@
+link:cncf-serverless-workflow-specification-support.html[CNCF Serverless Workflow Specification Support]
+and
+link:getting-familiar-with-our-tooling.html[Getting Familiar With Our Tooling].
+Also, you can go to the Cloud category from here to understand how to deploy your application in the cloud.
```
"Cloud category" should be a link

hbelmiro replied:
"Cloud" category is not a page. So, it doesn't have a link.

To understand what I'm talking about, see this example from the Antora docs:
[This](https://docs.antora.org/antora/latest/navigation/files-and-lists/) is a category and a page. So, it has a link. But the category above that (Navigation) is only a category, which doesn't have a link.

We can turn "Cloud" into a page and add links to the guides under this category to this page. But we'll end up with two places to update when adding a new guide (this page and the navigation file). This might not be a good idea.
Any thoughts on this @ricardozanini and @akumar074?

ibek replied:
Oh, right. Once we have a website we might have a link like Quarkus guides have: "https://quarkus.io/guides/#cloud"

ibek reviewed (changes requested):
Great content @hbelmiro , thank you for writing it up. As it is the first guide I have several suggestions which I believe would greatly help the first time users.

hbelmiro commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,257 @@
+
+This guide covered the creation of a Serverless Workflow application using Kogito.
+However, there is much more. We recommend continuing the journey with the getting started guides
+link:cncf-serverless-workflow-specification-support.html[CNCF Serverless Workflow Specification Support]
```
Good idea to add "Orchestration of OpenAPI based services".
Before removing "CNCF Serverless Workflow Specification Support", I just want to confirm, did you notice that this is not a link to the spec? It's a link to another guide under the getting-started category.

ibek replied:
I think that if a developer is looking for some feature / capability / use case they will first look at the available guides [1] and they should usually find the right guide there. Going into the details of the spec even if the CNCF guide / reference is an overview it is unnecessary and might be seen as too advanced content from my point of view. After the json in "=== Creating Your First Workflow" chapter it can be mentioned: "Did you know that the workflow definition is standard-based CNCF Serverless workflow?" and put there the link to guide for example.

[1] Understanding JQ expressions, Understanding workflow error handling, Handling events on workflows, etc.

ibek commented:
@hbelmiro Thank you for making the changes in the guide. The description of workflow is much better than I was hoping for. Great work there. I might have confused you when I was giving the examples of Understanding JQ expressions, Understanding workflow error handling and Handling events on workflows - I didn't meant to add them to the list of What's next items. I think "Getting Familiar With Our Tooling" and "Orchestration of OpenAPI Based Services" should be enough. Thank you

krisv reviewed (approved):
Left a few comments for potential improvements, but I'm fine if we decide to merge now and come back to them later if we want to as well.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,257 @@
+.Using Maven
+[source,shell]
+----
+mvn io.quarkus.platform:quarkus-maven-plugin:2.9.1.Final:create \
```
Oh, this should've been updated by his latest commit. We have a way to declare variables in the docs. So `quarkus_version` is one of them.

hbelmiro replied:
Yes. We're using a variable now.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,257 @@
+
+Hit `CTRL+C` to stop the application, or keep it running and enjoy the blazing fast hot-reload.
+
+== Testing
```
I guess we can redirect the testing content to this guide: https://github.com/kiegroup/kogito-docs/blob/main/serverlessworkflow/modules/ROOT/pages/testing-and-troubleshooting/basic-integration-tests-with-restassured.adoc

ricardozanini replied:
As soon as we incorporate this, we should wait for @hmanwani-rh's approval

hbelmiro replied:
Done!

hbelmiro commented:
> @hbelmiro Thank you for making the changes in the guide. The description of workflow is much better than I was hoping for. Great work there. I might have confused you when I was giving the examples of Understanding JQ expressions, Understanding workflow error handling and Handling events on workflows - I didn't meant to add them to the list of What's next items. I think "Getting Familiar With Our Tooling" and "Orchestration of OpenAPI Based Services" should be enough. Thank you

Many thanks for your feedback @ibek.
I removed those items from "What's next" and updated the guide with your other suggestions.

ibek commented:
@hbelmiro Thank you. Could you please add the changes/commits to the PR so I could check them? I will then approve the PR.

hbelmiro commented:
> @hbelmiro Thank you. Could you please add the changes/commits to the PR so I could check them? I will then approve the PR.

Sorry. I thought I had done.
Pushed now. Thanks.

ibek commented:
np and thank you @hbelmiro  It looks great

ricardozanini reviewed (approved):
LGTM as soon as @ibek suggestions are incorporated. Great work!

hbelmiro commented:
Thanks for the review @krisv.
I pushed the suggested changes.

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,286 @@
+installed (Optional)
+
+It's recommended that you take a look at
+link:getting-familiar-with-our-tooling.html[this guide] to get familiar with our tooling and the required dependencies.
```
@hbelmiro I think we can do this alternatively:

```suggestion
link:getting-familiar-with-our-tooling.html[this guide] to get familiar with our tooling and the required dependencies.
```
```suggestion

xref:getting-started/getting-familiar-with-our-tooling.adoc[this guide] to get familiar with our tooling and the required dependencies.
```

@akumar074 can you confirm?

ricardozanini replied:
@hbelmiro I think we can do this alternatively:

```suggestion
link:getting-familiar-with-our-tooling.html[this guide] to get familiar with our tooling and the required dependencies.
```
```suggestion

xref:getting-started/getting-familiar-with-our-tooling.adoc[this guide] to get familiar with our tooling and the required dependencies.
```

@akumar074 can you confirm?

ricardozanini commented:
@akumar074 @amadhusu can we safely ignore these CI errors?

krisv reviewed (approved):
Thanks @hbelmiro !

spolti reviewed (commented):
we can maybe add the package-lock.json on gitignore if it is not needed.

akumar074 commented:
@hbelmiro This PRs branch does not have updated `.github/workflows/sw-docs-pr.yml` 
Please compare 
[1] https://github.com/hbelmiro/kogito-docs/blob/8a94411aede305e7ede4b53c904ded876247fd16/.github/workflows/sw-docs-pr.yml#L24
and
[2] https://github.com/kiegroup/kogito-docs/blob/e767491702eb4bcf4160211e6b1bd307572e3432/.github/workflows/sw-docs-pr.yml#L24

akumar074 commented:
@hbelmiro This branch does not have an updated `package.json` file. It is missing the `scripts` key. You can compare it with the `upstream/main` branch.
![Screenshot from 2022-06-03 09-19-39](https://user-images.githubusercontent.com/19465051/171782470-0c116af0-e82a-4f4a-bb24-99f772bac07f.png)

hbelmiro commented:
> we can maybe add the package-lock.json on gitignore if it is not needed.

+1

@akumar074 WDYT?

akumar074 commented:
> > we can maybe add the package-lock.json on gitignore if it is not needed.
> 
> +1
> 
> @akumar074 WDYT?

@hbelmiro @spolti I think we should commit package-lock.json. package-lock.json serves a good purpose to keep dependency in check. Right now we just have 3 dependencies, but it may later increase.

hmanwani-rh reviewed (commented):
@hbelmiro @ricardozanini I've added changes to the content. Please take a look and let me know if you're good with the changes.

hbelmiro commented:
> @hbelmiro @ricardozanini I've added changes to the content. Please take a look and let me know if you're good with the changes.

Thanks for your improvements, @hmanwani-rh.
Looks great!

ricardozanini commented on serverlessworkflow/modules/ROOT/pages/getting-started/create-your-first-workflow-service.adoc:
```diff
@@ -1 +1,282 @@
+* <<running-application,Running your Serverless Workflow application>>
+* <<testing-application,Testing your Serverless Workflow application>>
+
+Also, you can directly access an example application by cloning `kogito-quarkus-examples/serverless-workflow-hello-world` from the https://github.com/hbelmiro/kogito-examples/tree/main/kogito-quarkus-examples/serverless-workflow-hello-world[GitHub repository].
```
That's fine using this link now, but it's likely to be changed.

ricardozanini commented:
@hmanwani-rh it's ok, you can merge this PR.
